File: .claude/hooks/session-dedup-autodisable.mjs
#!/usr/bin/env node
// @hookstack stop-session-dedup-autodisable
// Auto-désactive les hooks Stop qui ont échoué ≥ N fois de suite (Stop)
//
// Contrat partagé : un hook Stop qui veut bénéficier du watchdog incrémente
// /tmp/claude-hook-counters/<slug>.counter à chaque échec (et le supprime en cas
// de succès), puis vérifie l'absence de <slug>.disabled avant de s'exécuter.
// Ce hook pose le marqueur .disabled dès que le compteur atteint MAX_FAILURES.
import { readFileSync, existsSync, readdirSync, writeFileSync } from 'fs';
import { join } from 'path';
import { fileURLToPath } from 'url';

const MAX_FAILURES = 3;

export function run({
  exists = existsSync,
  readdir = readdirSync,
  readFile = readFileSync,
  writeFile = writeFileSync,
  counterDir = '/tmp/claude-hook-counters',
} = {}) {
  if (!exists(counterDir)) return null;

  try {
    const counters = readdir(counterDir).filter((f) => f.endsWith('.counter'));
    const disabled = [];
    for (const f of counters) {
      let count = 0;
      try {
        count = parseInt(readFile(join(counterDir, f), 'utf8').trim(), 10) || 0;
      } catch {
        continue;
      }
      if (count < MAX_FAILURES) continue;
      const slug = f.replace('.counter', '');
      const marker = join(counterDir, `${slug}.disabled`);
      if (!exists(marker)) writeFile(marker, '');
      disabled.push(slug);
    }

    if (!disabled.length) return null;

    const message =
      `[session-dedup] Hooks désactivés (${MAX_FAILURES}+ échecs) : ${disabled.join(', ')}\n` +
      `[session-dedup] Supprimez ${counterDir}/<slug>.disabled (et .counter) pour réactiver.\n`;
    return { disabled, message };
  } catch {
    // Erreur de lecture — ignorer silencieusement
    return null;
  }
}

/* v8 ignore next 4 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const result = run();
  if (result?.message) process.stderr.write(result.message);
}

File: .claude/hooks/session-dedup-autodisable.py
#!/usr/bin/env python3
# @hookstack stop-session-dedup-autodisable
"""Auto-disables Stop hooks that failed >= N times in a row (Stop).

Shared contract: a Stop hook that wants the watchdog increments
/tmp/claude-hook-counters/<slug>.counter on each failure (and removes it on
success), then checks for the absence of <slug>.disabled before running.
This hook places the .disabled marker as soon as the counter reaches MAX_FAILURES.
"""
import os
import sys

MAX_FAILURES = 3


def run(
    *,
    exists=None,
    readdir=None,
    read_file=None,
    write_file=None,
    counter_dir="/tmp/claude-hook-counters",
):
    if exists is None:
        exists = os.path.exists
    if readdir is None:
        readdir = os.listdir
    if read_file is None:
        read_file = lambda p: open(p, "r", encoding="utf8").read()
    if write_file is None:
        write_file = lambda p, content: open(p, "w", encoding="utf8").write(content)

    if not exists(counter_dir):
        return None

    try:
        counters = [f for f in readdir(counter_dir) if f.endswith(".counter")]
        disabled = []
        for f in counters:
            try:
                count = int(read_file(os.path.join(counter_dir, f)).strip()) or 0
            except Exception:
                continue
            if count < MAX_FAILURES:
                continue
            slug = f.replace(".counter", "")
            marker = os.path.join(counter_dir, f"{slug}.disabled")
            if not exists(marker):
                write_file(marker, "")
            disabled.append(slug)

        if not disabled:
            return None

        message = (
            f"[session-dedup] Hooks désactivés ({MAX_FAILURES}+ échecs) : {', '.join(disabled)}\n"
            f"[session-dedup] Supprimez {counter_dir}/<slug>.disabled (et .counter) pour réactiver.\n"
        )
        return {"disabled": disabled, "message": message}
    except Exception:
        # Erreur de lecture — ignorer silencieusement
        return None


if __name__ == "__main__":
    result = run()
    if result and result.get("message"):
        sys.stderr.write(result["message"])
